x-pack/plugins/security/server/lib/authentication/providers/token.js

```diff
@@ -0,0 +1,215 @@
+    }
+
+    // if we still can't attempt auth, try authenticating via state (session token)
+    if (authenticationResult.notHandled()) {
+      authenticationResult = await this._authenticateViaState(request, state);
+      if (authenticationResult.failed() && isAccessTokenExpiredError(authenticationResult.error)) {
+        authenticationResult = this._authenticateViaRefreshToken(request, state);
+      }
+    }
+
+    // finally, if authentication still can not be handled for this
+    // request/state combination, redirect to the login page if appropriate
+    if (authenticationResult.notHandled() && canRedirectRequest(request)) {
+      const nextURL = encodeURIComponent(`${this._options.basePath}${request.url.path}`);
+      return AuthenticationResult.redirectTo(
+        `${this._options.basePath}/login?next=${nextURL}`
+      );
+    }
+
+    return authenticationResult;
+  }
+
+  /**
+   * Redirects user to the login page preserving query string parameters.
+   * @param {Hapi.Request} request HapiJS request instance.
+   * @returns {Promise.<DeauthenticationResult>}
+   */
+  async deauthenticate(request) {
+    // Query string may contain the path where logout has been called or
+    // logout reason that login page may need to know.
+    return DeauthenticationResult.redirectTo(
+      `${this._options.basePath}/login${request.url.search}`
+    );
+  }
+
+  /**
+   * Validates whether request contains `Basic ***` Authorization header and just passes it
+   * forward to Elasticsearch backend.
+   * @param {Hapi.Request} request HapiJS request instance.
+   * @returns {Promise.<AuthenticationResult>}
+   * @private
+   */
+  async _authenticateViaHeader(request) {
+    this._options.log(['debug', 'security', 'token'], 'Trying to authenticate via header.');
+
+    const authorization = request.headers.authorization;
+    if (!authorization) {
+      this._options.log(['debug', 'security', 'token'], 'Authorization header is not presented.');
+      return AuthenticationResult.notHandled();
+    }
+
+    const authenticationSchema = authorization.split(/\s+/)[0];
+    if (authenticationSchema.toLowerCase() !== 'basic') {
+      this._options.log(['debug', 'security', 'token'], `Unsupported authentication schema: ${authenticationSchema}`);
+
+      // It's essential that we fail if non-empty, but unsupported authentication schema
+      // is provided to allow authenticator to consult other authentication providers
+      // that may support that schema.
+      return AuthenticationResult.failed(
+        Boom.badRequest(`Unsupported authentication schema: ${authenticationSchema}`)
+      );
+    }
+
+    try {
+      const user = await this._options.client.callWithRequest(request, 'shield.authenticate');
+
+      this._options.log(['debug', 'security', 'token'], 'Request has been authenticated via header.');
+
+      return AuthenticationResult.succeeded(user, { authorization });
```

Comment:
I'll have to double check, but I think passing `user` to `succeeded` isn't the problem here as that shouldn't include the credentials (at least not the password). But I think passing `authorization` here will have the effect you describe, so we definitely don't want that.

Reply:
> I'll have to double check, but I think passing `user` to `succeeded` isn't the problem here as that shouldn't include the credentials (at least not the password)

It shouldn't be a problem, and moreover it's required for `succeeded` and it's not used for session.

>  which can't be invalidated with the logout. We'll likely want to perform the same flow as the "authenticate via login attempt"

Hmm, correct me if I'm wrong, but if it's in cookie, the `deauthenticate` that is called on logout should clear it as well. But having said that I agree that we don't want to store `Bearer xxxx` in the cookie when we authenticate user via `header` (and should remove this from "feature" from basic provider `auth via header` as well since now we have `login attempt` thing), but if we don't store it in the session cookie during "login attempt", where do we get access/refresh tokens from for the next request?

Reply:
Hmm, I was talking about two cases:

1. `_authenticateViaHeader` - here we expect to receive `Authorization: Bearer xxxx` and hence we shouldn't store anything in the cookie since we expect every request to have this header (e.g. they have some OAuth proxy in front of Kibana that supplies it). Currently basic auth provider **stores** it in the cookie, but with `loginAttempt` we can finally fix that.

2. `_authenticateViaLoginAttempt` - here we expect to receive credentials that we exchange for the token pair. It's probably one time action (e.g. initiated from the login page) so we should store token pair in the cookie (and "forget" credentials) and allow consequent requests to rely on generated tokens from the cookie instead.

Or I'm misunderstanding something? :)

Reply:
I believe we're on the same page, I forgot that we could see either the `Authorization: Bearer xxxx` or the `Authorization: Basic xxxx` headers when using the `_authenticateViaHeader` method, but I believe it makes sense to not store these in the cookie in either situation.

Reply:
> I believe we're on the same page

Great! 

> , but I believe it makes sense to not store these in the cookie in either situation.

Yes, that's where `loginAttempt` that we didn't have before can help us.